pkg/planner/cardinality/row_count_column.go

```diff
@@ -177,7 +179,15 @@ func equalRowCountOnColumn(sctx planctx.PlanContext, c *statistics.Column, val t
 		if modifyCount == 0 {
 			return 0, nil
 		}
-		return 1, nil
+		// Reset to the original NDV, or if no NDV - derive an NDV using sqrt
+		if c.Histogram.NDV > 0 {
+			histNDV = float64(c.Histogram.NDV)
+		} else {
+			histNDV = math.Sqrt(max(c.Histogram.NotNullCount(), float64(modifyCount)))
+		}
+		// Return the min of the original notNullCount or the modifyCount/NDV. This is to reduce the
+		// risk of too large or too small an estimate if modifyCount is large or NotNullCount is small
+		return max(1, min(c.Histogram.NotNullCount(), float64(modifyCount)/histNDV)), nil
```

Comment:
I think you should use `c.NotNullCount` here.

Reply:
We should use the TopN count + Hist Count here. The c.Histogram.NotNullCount() just for hist.

Reply:
I have corrected this in latest revision